### ISSUE TITLE
@index directive with multiple sortKeys is not filtered by second field

### ISSUE THREAD

**Issue by armanian** [aws-amplify/amplify-category-api#802]
### Before opening, please confirm:

- [X] I have installed the latest version of the Amplify CLI (see above), and confirmed that the issue still persists.
- [X] I have [searched for duplicate or closed issues](https://github.com/aws-amplify/amplify-cli/issues?q=is%3Aissue+).
- [X] I have read the guide for [submitting bug reports](https://github.com/aws-amplify/amplify-cli/blob/master/CONTRIBUTING.md#bug-reports).
- [X] I have done my best to include a minimal, self-contained set of instructions for consistently reproducing the issue.
- [X] I have removed any sensitive information from my code snippets and submission.

### How did you install the Amplify CLI?

npm install -g @aws-amplify/cli

### If applicable, what version of Node.js are you using?

v16.17.0

### Amplify CLI Version

10.0.0 & 9.2.1

### What operating system are you using?

MacOS 12.6

### Did you make any manual changes to the cloud resources managed by Amplify? Please describe the changes made.

No manual changes made

### Amplify Categories

api

### Amplify Commands

push

### Describe the bug

As an example scenario: I am trying to filter table `Order` by Int field `updated` and sort by Int field `created`.  For this I have the index `listOrdersByCreated` with two sortKeys `["created", "updated"]`.

```graphql
enum OrderType {
  HARDWARE
  SOFTWARE
}

type Order @model @auth(rules: [{ allow: private }]) {
  id: ID!
  type: OrderType!
    @index(name: "byCreated", queryField: "listOrdersByCreated", sortKeyFields: ["created", "updated"])
  created: Int!
  updated: Int!
}
```

For filtering I use `between`. The `created` field is filtered, but the `updated` field is ignored:

```graphql
query MyQuery {
  listOrdersByCreated(
    type: HARDWARE
    createdUpdated: {
      between: [
        { created: 0, updated: 0 },
        { created: 2003187334, updated: 2003187334 }
      ]
    }
  ) {
    items {
      created
      updated
    }
  }
}
```
Result (correct):
```json
{
  "data": {
    "listOrdersByCreated": {
      "items": [
        {
          "created": 1663321326,
          "updated": 1664421326
        },
        {
          "created": 1663321333,
          "updated": 1664421333
        },
        {
          "created": 1663321339,
          "updated": 1664421339
        }
      ]
    }
  }
}
```
filter by created:
```graphql
between: [
  { created: 0, updated: 0 },
  { created: 1, updated: 2003187334 }
]
```
Result (correct):
```json
{
  "data": {
    "listOrdersByCreated": {
      "items": []
    }
  }
}
```
filter by updated:
```graphql
between: [
  { created: 0, updated: 0 },
  { created: 2003187334, updated: 1 }
]
```
Result (wrong):
```json
{
  "data": {
    "listOrdersByCreated": {
      "items": [
        {
          "created": 1663321326,
          "updated": 1664421326
        },
        {
          "created": 1663321333,
          "updated": 1664421333
        },
        {
          "created": 1663321339,
          "updated": 1664421339
        }
      ]
    }
  }
}
```
Apparently, the second field is completely ignored.

### Expected behavior

I can filter my index by `updated` too.

### Reproduction steps

1. Add api & use schema
2. Make call in AppSync

### GraphQL schema(s)

<details>

```graphql
enum OrderType {
  HARDWARE
  SOFTWARE
}

type Order @model @auth(rules: [{ allow: private }]) {
  id: ID!
  type: OrderType!
    @index(name: "byGeo", queryField: "listOrdersByCreated", sortKeyFields: ["created", "updated"])
  created: Int!
  updated: Int!
}
```

</details>

### Log output

<details>

```
# Put your logs below this line


```

</details>

**Comment by armanian** [comment]
Any updates? The bug is very critical.

**Comment by jcollinc** [comment]
I was able to reproduce with the steps provided, currently looking into this

**Comment by armanian** [comment]
@jcollinc Could you solve the problem?

**Comment by armanian** [comment]
@josefaidt any updates?

**Comment by armanian** [comment]
Does anyone have an idea how to solve this? The problem still exists.

**Comment by iartemiev** [comment]
@armanian apologies for the slow response on this. I'll provide a workaround first and an explanation for the behavior you're seeing below.

tl;dr:
What you're seeing is expected behavior (see explanation below). You can retrieve the result you're looking for by splitting up your predicate into a query expression + filter expression like this:

```gql
query MyQuery {
  listOrdersByCreated(
    type: HARDWARE
    createdUpdated: {
      between: [
        { created: 0 },
        { created: 2003187334 }
      ]
    },
    filter: {
      updated: {
        between: [0, 1]
      }
    }
  ) {
    items {
      created
      updated
    }
  }
}
```

#### Explanation for current behavior

It's actually not possible to include both of these predicates in the query expression and get the result you're expecting due to the way range queries work in DynamoDB. A composite sort key like what you've defined is stored in a single attribute with a `#` separator in DDB.

Example:
| type  | created#updated | created | updated |
| ------------- | ------------- | ------------- | ------------- |
| HARDWARE  | 1663321326#1664421326  | 1663321326 | 1664421326 |
| HARDWARE  | 1663321333#1664421333  | 1663321333 | 1664421333 |
| HARDWARE  | 1663321339#1664421339  | 1663321339 | 1664421339 |

DynamoDB does not provide a way to evaluate both sides of this value separately in a conjunction.

Given these GraphQL parameters
```
type: HARDWARE,
createdUpdated: {
  between: [
    { created: 0, updated: 0 },
    { created: 2003187334, updated: 1 }
  ]
}
```
The resolvers for your index query produce the following query expression against the Order table:
```
{
  "query": {
    "expression": "#type = :type AND #sortKey BETWEEN :sortKey0 AND :sortKey1",
    "expressionNames": {
      "#type": "type",
      "#sortKey": "created#updated"
    },
    "expressionValues": {
      ":type": {
        "S": "SOFTWARE"
      },
      ":sortKey0": {
        "S": "0#0"
      },
      ":sortKey1": {
        "S": "2003187334#1"
      }
    }
  },
  "index": "byCreated",
  "scanIndexForward": true
}
```

When evaluating the expression, DDB performs a lexicographic string comparison. In other words:

Return all values that are between `'0#0'` and `'2003187334#1'`. All the values in the table above fall within that range.

We can verify in JS, which also compares strings lexicographically:
```js
const start = '0#0';
const end = '2003187334#1';

const keys = [
  '1663321326#1664421326',
  '1663321333#1664421333',
  '1663321339#1664421339'
]

keys.forEach(key => console.log(key > start && key < end)) // true, true, true
```

**Comment by armanian** [comment]
@iartemiev Thank you for the message.

How can such a filter be implemented? Is there a solution for this problem without having to use the cost-intensive solution such as "ElasticSearch"?

We have the same problem with the geohash filter.

**Comment by iartemiev** [comment]
> How can such a filter be implemented? 

@armanian if you're using AWS Amplify, you would pass the filter to `API.graphql` along with the other variables, e.g.
```ts
const result = await client.graphql({
  query: listOrdersByCreated,
  variables: {
    type: OrderType.HARDWARE,
    createdUpdated: {
      between: [
        { created: 0 },
        { created: 2003187334 }
      ]
    },
    filter: {
      updated: {
        between: [0, 1]
      }
    }
  }
});
```
